File: convert_json_to_txt.py
import json
import os
import re
from collections import defaultdict

# Transliteration mapping for Kazakh Cyrillic to clean Latin
TRANSLIT_MAP = {
    'а': 'a', 'ә': 'a', 'б': 'b', 'в': 'v', 'г': 'g', 'ғ': 'g', 'д': 'd', 'е': 'e', 'ё': 'yo', 'ж': 'zh',
    'з': 'z', 'и': 'i', 'й': 'y', 'к': 'k', 'қ': 'q', 'л': 'l', 'м': 'm', 'н': 'n', 'ң': 'n', 'о': 'o',
    'ө': 'o', 'п': 'p', 'р': 'r', 'с': 's', 'т': 't', 'у': 'u', 'ұ': 'u', 'ү': 'u', 'ф': 'f', 'х': 'kh',
    'һ': 'h', 'ц': 'ts', 'ч': 'ch', 'ш': 'sh', 'щ': 'shch', 'ы': 'y', 'і': 'i', 'э': 'e', 'ю': 'yu', 'я': 'ya',
}

def slugify(text):
    if not text:
        return ""
    text = text.lower()
    res = []
    for char in text:
        if char in TRANSLIT_MAP:
            res.append(TRANSLIT_MAP[char])
        elif char.isalnum():
            res.append(char)
        elif char in (' ', '-', '_'):
            res.append('-')
    slug = "".join(res)
    # clean up double hyphens
    slug = re.sub(r'-+', '-', slug)
    return slug.strip('-')

def main():
    base_dir = os.path.dirname(os.path.abspath(__file__))
    json_path = os.path.join(base_dir, "src", "data", "family.json")
    txt_path = os.path.join(base_dir, "family.txt")

    if not os.path.exists(json_path):
        print(f"Error: {json_path} not found.")
        return

    with open(json_path, "r", encoding="utf-8") as f:
        data = json.load(f)

    by_old_id = {node["id"]: node for node in data}

    # Generate readable slug IDs mapping: old_id -> new_slug
    old_to_new_id = {}
    generated_slugs = set()

    for node in data:
        name_slug = slugify(node.get("name", ""))
        
        father_node = None
        if "fatherId" in node and node["fatherId"]:
            father_node = by_old_id.get(node["fatherId"])
            
        spouse_node = None
        if "spouseOf" in node and node["spouseOf"]:
            spouse_node = by_old_id.get(node["spouseOf"])

        if father_node:
            father_name_slug = slugify(father_node.get("name", ""))
            base_slug = f"{name_slug}-{father_name_slug}"
        elif spouse_node:
            spouse_name_slug = slugify(spouse_node.get("name", ""))
            base_slug = f"{name_slug}-{spouse_name_slug}"
        else:
            base_slug = name_slug

        # Ensure uniqueness
        candidate = base_slug
        counter = 2
        while candidate in generated_slugs:
            candidate = f"{base_slug}-{counter}"
            counter += 1
        
        generated_slugs.add(candidate)
        old_to_new_id[node["id"]] = candidate

    # Update relationships to point to new slug IDs
    updated_data = []
    for node in data:
        new_node = node.copy()
        new_node["id"] = old_to_new_id[node["id"]]
        if "fatherId" in node and node["fatherId"]:
            new_node["fatherId"] = old_to_new_id.get(node["fatherId"])
        if "motherId" in node and node["motherId"]:
            new_node["motherId"] = old_to_new_id.get(node["motherId"])
        if "spouseOf" in node and node["spouseOf"]:
            new_node["spouseOf"] = old_to_new_id.get(node["spouseOf"])
        updated_data.append(new_node)

    by_new_id = {node["id"]: node for node in updated_data}

    # Identify children by father
    children_by_father = defaultdict(list)
    for node in updated_data:
        if node.get("fatherId"):
            children_by_father[node["fatherId"]].append(node)

    # Identify inline spouses: female spouses who have no father in the database
    # These will be printed on the same line as their husband.
    inline_spouses_by_husband = defaultdict(list)
    written_inline_spouses = set()

    for node in updated_data:
        if node.get("spouseOf") and not node.get("fatherId"):
            inline_spouses_by_husband[node["spouseOf"]].append(node)
            written_inline_spouses.add(node["id"])

    # Find root nodes: nodes with no father and not written inline
    roots = []
    for node in updated_data:
        if not node.get("fatherId") and node["id"] not in written_inline_spouses:
            roots.append(node)

    # Format attributes string
    def format_attrs(node, exclude_keys=None):
        if exclude_keys is None:
            exclude_keys = set()
        
        # Determine default ID
        name_slug = slugify(node.get("name", ""))
        father_node = by_new_id.get(node.get("fatherId")) if node.get("fatherId") else None
        spouse_node = by_new_id.get(node.get("spouseOf")) if node.get("spouseOf") else None
        
        if father_node:
            default_id = f"{name_slug}-{slugify(father_node.get('name', ''))}"
        elif spouse_node:
            default_id = f"{name_slug}-{slugify(spouse_node.get('name', ''))}"
        else:
            default_id = name_slug
            
        parts = []
        # Only output id if it deviates from the default slug pattern
        if node["id"] != default_id and "id" not in exclude_keys:
            parts.append(f"id: {node['id']}")
            
        for key in ["lastName", "gender", "birthday", "motherId", "spouseOf", "notes"]:
            if key in exclude_keys:
                continue
            val = node.get(key)
            if val is not None and val != "":
                # Clean notes of newlines
                if key == "notes":
                    val = val.replace("\n", " ").replace("\r", "")
                parts.append(f"{key}: {val}")
                
        if parts:
            return f" [{', '.join(parts)}]"
        return ""

    output_lines = []

    def write_node_tree(node, depth):
        indent = "  " * depth
        line = f"{indent}- {node['name']}"
        
        # Append attributes
        line += format_attrs(node)
        
        # Append inline spouses if any
        spouses = inline_spouses_by_husband.get(node["id"], [])
        for sp in spouses:
            spouse_str = f" (spouse: {sp['name']}"
            spouse_attrs = format_attrs(sp, exclude_keys={"spouseOf", "gender"})  # gender is implicitly female
            spouse_str += spouse_attrs + ")"
            line += spouse_str
            
        output_lines.append(line)
        
        # Recursively write children
        children = children_by_father.get(node["id"], [])
        # Sort children by birthday or index to keep consistent structure
        children.sort(key=lambda x: x.get("birthday", ""))
        for child in children:
            write_node_tree(child, depth + 1)

    # Sort roots: put the main shezhire root first (usually Кіші жүз or the one with most descendants)
    # To identify the main tree, we can count the size of each tree
    descendant_counts = {}
    def count_descendants(nid):
        if nid in descendant_counts:
            return descendant_counts[nid]
        count = len(children_by_father[nid])
        for child in children_by_father[nid]:
            count += count_descendants(child["id"])
        descendant_counts[nid] = count
        return count

    roots.sort(key=lambda r: count_descendants(r["id"]), reverse=True)

    for root in roots:
        write_node_tree(root, 0)

    with open(txt_path, "w", encoding="utf-8") as f:
        f.write("\n".join(output_lines) + "\n")

    print(f"Successfully converted {len(data)} nodes to {txt_path}")
    print(f"Mapped {len(old_to_new_id)} old IDs to readable slugs.")

if __name__ == "__main__":
    main()

File: scripts/convert_json_to_txt.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Transliteration mapping for Kazakh Cyrillic to clean Latin
const TRANSLIT_MAP = {
  'а': 'a', 'ә': 'a', 'б': 'b', 'в': 'v', 'г': 'g', 'ғ': 'g', 'д': 'd', 'е': 'e', 'ё': 'yo', 'ж': 'zh',
  'з': 'z', 'и': 'i', 'й': 'y', 'к': 'k', 'қ': 'q', 'л': 'l', 'м': 'm', 'н': 'n', 'ң': 'n', 'о': 'o',
  'ө': 'o', 'п': 'p', 'р': 'r', 'с': 's', 'т': 't', 'у': 'u', 'ұ': 'u', 'ү': 'u', 'ф': 'f', 'х': 'kh',
  'һ': 'h', 'ц': 'ts', 'ч': 'ch', 'ш': 'sh', 'щ': 'shch', 'ы': 'y', 'і': 'i', 'э': 'e', 'ю': 'yu', 'я': 'ya',
};

function slugify(text) {
  if (!text) return "";
  text = text.toLowerCase();
  const res = [];
  for (const char of text) {
    if (TRANSLIT_MAP[char] !== undefined) {
      res.push(TRANSLIT_MAP[char]);
    } else if (/[a-z0-9]/i.test(char)) {
      res.push(char);
    } else if (char === ' ' || char === '-' || char === '_') {
      res.push('-');
    }
  }
  let slug = res.join('');
  slug = slug.replace(/-+/g, '-');
  return slug.replace(/^-+|-+$/g, '');
}

function main() {
  const baseDir = path.dirname(__dirname);
  const jsonPath = path.join(baseDir, "src", "data", "family.json");
  const txtPath = path.join(baseDir, "family.txt");

  if (!fs.existsSync(jsonPath)) {
    console.error(`Error: ${jsonPath} not found.`);
    process.exit(1);
  }

  const rawData = fs.readFileSync(jsonPath, "utf-8");
  const data = JSON.parse(rawData);

  const byOldId = {};
  for (const node of data) {
    byOldId[node.id] = node;
  }

  // Generate readable slug IDs mapping: old_id -> new_slug
  const oldToNewId = {};
  const generatedSlugs = new Set();

  for (const node of data) {
    const nameSlug = slugify(node.name || "");
    
    let fatherNode = null;
    if (node.fatherId) {
      fatherNode = byOldId[node.fatherId];
    }
      
    let spouseNode = null;
    if (node.spouseOf) {
      spouseNode = byOldId[node.spouseOf];
    }

    let baseSlug = "";
    if (fatherNode) {
      const fatherNameSlug = slugify(fatherNode.name || "");
      baseSlug = `${nameSlug}-${fatherNameSlug}`;
    } else if (spouseNode) {
      const spouseNameSlug = slugify(spouseNode.name || "");
      baseSlug = `${nameSlug}-${spouseNameSlug}`;
    } else {
      baseSlug = nameSlug;
    }

    // Ensure uniqueness
    let candidate = baseSlug;
    let counter = 2;
    while (generatedSlugs.has(candidate)) {
      candidate = `${baseSlug}-${counter}`;
      counter++;
    }
    
    generatedSlugs.add(candidate);
    oldToNewId[node.id] = candidate;
  }

  // Update relationships to point to new slug IDs
  const updatedData = [];
  for (const node of data) {
    const newNode = { ...node };
    newNode.id = oldToNewId[node.id];
    if (node.fatherId) {
      newNode.fatherId = oldToNewId[node.fatherId];
    }
    if (node.motherId) {
      newNode.motherId = oldToNewId[node.motherId];
    }
    if (node.spouseOf) {
      newNode.spouseOf = oldToNewId[node.spouseOf];
    }
    updatedData.push(newNode);
  }

  const byNewId = {};
  for (const node of updatedData) {
    byNewId[node.id] = node;
  }

  // Identify children by father
  const childrenByFather = {};
  for (const node of updatedData) {
    if (node.fatherId) {
      if (!childrenByFather[node.fatherId]) {
        childrenByFather[node.fatherId] = [];
      }
      childrenByFather[node.fatherId].push(node);
    }
  }

  // Identify inline spouses: female spouses who have no father in the database
  const inlineSpousesByHusband = {};
  const writtenInlineSpouses = new Set();

  for (const node of updatedData) {
    if (node.spouseOf && !node.fatherId) {
      if (!inlineSpousesByHusband[node.spouseOf]) {
        inlineSpousesByHusband[node.spouseOf] = [];
      }
      inlineSpousesByHusband[node.spouseOf].push(node);
      writtenInlineSpouses.add(node.id);
    }
  }

  // Find root nodes
  const roots = [];
  for (const node of updatedData) {
    if (!node.fatherId && !writtenInlineSpouses.has(node.id)) {
      roots.push(node);
    }
  }

  // Format attributes string
  function formatAttrs(node, excludeKeys = new Set()) {
    const nameSlug = slugify(node.name || "");
    const fatherNode = node.fatherId ? byNewId[node.fatherId] : null;
    const spouseNode = node.spouseOf ? byNewId[node.spouseOf] : null;
    
    let defaultId = "";
    if (fatherNode) {
      defaultId = `${nameSlug}-${slugify(fatherNode.name || "")}`;
    } else if (spouseNode) {
      defaultId = `${nameSlug}-${slugify(spouseNode.name || "")}`;
    } else {
      defaultId = nameSlug;
    }
      
    const parts = [];
    if (node.id !== defaultId && !excludeKeys.has("id")) {
      parts.push(`id: ${node.id}`);
    }
      
    const keys = ["lastName", "gender", "birthday", "motherId", "spouseOf", "notes"];
    for (const key of keys) {
      if (excludeKeys.has(key)) continue;
      
      // Skip redundant motherId (when the mother is the spouse of the father)
      if (key === "motherId" && node.motherId && node.fatherId) {
        const motherNode = byNewId[node.motherId];
        if (motherNode && motherNode.spouseOf === node.fatherId) {
          continue;
        }
      }

      let val = node[key];
      if (val !== undefined && val !== null && val !== "") {
        if (key === "notes") {
          val = val.replace(/\n/g, " ").replace(/\r/g, "");
        }
        parts.push(`${key}: ${val}`);
      }
    }
          
    if (parts.length > 0) {
      return ` [${parts.join(", ")}]`;
    }
    return "";
  }

  const outputLines = [];

  function writeNodeTree(node, depth) {
    const indent = "  ".repeat(depth);
    let line = `${indent}- ${node.name}`;
    
    // Append attributes
    line += formatAttrs(node);
    
    // Append inline spouses
    const spouses = inlineSpousesByHusband[node.id] || [];
    for (const sp of spouses) {
      let spouseStr = ` (spouse: ${sp.name}`;
      const spouseAttrs = formatAttrs(sp, new Set(["spouseOf", "gender"]));
      spouseStr += spouseAttrs + ")";
      line += spouseStr;
    }
      
    outputLines.push(line);
    
    // Recursively write children
    const children = childrenByFather[node.id] || [];
    children.sort((a, b) => (a.birthday || "").localeCompare(b.birthday || ""));
    for (const child of children) {
      writeNodeTree(child, depth + 1);
    }
  }

  // Sort roots by descendant count descending
  const descendantCounts = {};
  function countDescendants(nid) {
    if (descendantCounts[nid] !== undefined) {
      return descendantCounts[nid];
    }
    const children = childrenByFather[nid] || [];
    let count = children.length;
    for (const child of children) {
      count += countDescendants(child.id);
    }
    descendantCounts[nid] = count;
    return count;
  }

  roots.sort((a, b) => countDescendants(b.id) - countDescendants(a.id));

  for (const root of roots) {
    writeNodeTree(root, 0);
  }

  fs.writeFileSync(txtPath, outputLines.join("\n") + "\n", "utf-8");
  console.log(`Successfully converted ${data.length} nodes to ${txtPath}`);
}

main();
